### PR TITLE
Dereference NULL return value

### PR DESCRIPTION
**Pull request by kansal**
No check on the null value of metadatafile. If null, the following operations failed. Added null check for the metadatafile.

### CODE REVIEW

File: plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java [hunk 1]
```diff
@@ -368,8 +368,14 @@ private boolean setup() {
                 _idpMetaDataProvider = new HTTPMetadataProvider(_timer, client, idpMetaDataUrl);
             } else {
                 File metadataFile = PropertiesUtil.findConfigFile(idpMetaDataUrl);
-                s_logger.debug("Provided Metadata is not a URL, trying to read metadata file from local path: " + metadataFile.getAbsolutePath());
-                _idpMetaDataProvider = new FilesystemMetadataProvider(_timer, metadataFile);
+                if (metadataFile == null) {
+                    s_logger.error("Metadata file returned null");
```
Comment (kishankavala) [inline plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java]: Can you please update the error message: "Provided Metadata is not a URL, Unable to locate metadata file from local path"

Comment (karuturi) [inline plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java]: Can you also add the file path its trying to fing (idpMetaDataUrl) to the error message?

**Comment by DaanHoogland** [comment]
thanks @kansal
Can you add some test cases (for null values found for instance)?
LGTM

**Comment by kansal** [comment]
@kishankavala: The message has been changed.
@karuturi: Added the path. Please check

I have also made another change with respect to solving this problem of dereferencing of null pointer. Kindly review the changes.

**Comment by karuturi** [comment]
instead of a notnull check everywhere, can you use a Non Nulllable ArrayList for tags or use CollectionUtils.addIgnoreNull() api?

**Comment by kansal** [comment]
@DaanHoogland Added the test file. Checking only the false cases. @wilderrodrigues Got away with the IF statements and replaced with the api @karuturi suggested. Please review the changes.

**Comment by kishankavala** [comment]
LGTM. @bhaisaab can you also take a look at the changes?

**Comment by wilderrodrigues** [comment]
LGTM :+1: 

Thanks for the improvements!

File: plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java [hunk 2]
```diff
@@ -119,12 +123,12 @@
     @Override
     public boolean start() {
         if (isSAMLPluginEnabled()) {
-            setup();
             s_logger.info("SAML auth plugin loaded");
+            return setup();
         } else {
```
Comment (bhaisaab) [inline plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java]: we can remove else {} here just keep;
if (condition){
return stuff();
}
return otherstuff();

File: plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java [hunk 3]
```diff
@@ -358,7 +363,8 @@ private boolean setup() {
         }
         _timer = new Timer();
         final HttpClient client = new HttpClient();
-        final String idpMetaDataUrl = SAMLIdentityProviderMetadataURL.value();
+        //final String idpMetaDataUrl = SAMLIdentityProviderMetadataURL.value();
```
Comment (bhaisaab) [inline plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java]: Remove commented deadcode, or simply keep final String idpMetaDataUrl = SAMLIdentityProviderMetadataURL.value(); as not many consumers of the new getSAMLIdentityProviderMetadataURL() method.

**Comment by bhaisaab** [comment]
@kansal Please see my comments. While it's okay, please squash the commits into two separate commits; one targeting the saml related stuff and second one targeting the other component, so as to allow us to port the commits to other branches.

File: plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java (hunk 2 above)
Comment (kansal) [inline plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java]: @bhaisaab I feel we can go with the if-else style because it anyhow is doing the same thing plus it give better code readability and logging. Your take?

File: plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java (hunk 3 above)
Comment (kansal) [inline plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAML2AuthManagerImpl.java]: @bhaisaab  Sure. Removing the comment. The function call is needed for mocking at the time of unit testing.

**Comment by kansal** [comment]
Closing this pull request. Will open one for each issue.
